### PR TITLE
Feat/tfn 791 search show operators

### PR DESCRIPTION
**Pull request by tejpowar**
# Description

Page setup and search functionality to find operators

# Testing instructions

Browse to http://localhost:5555/searchOperators

# Type of change

-   [x] feat - A new feature
-   [ ] fix - A bug fix
-   [ ] docs - Documentation only changes
-   [ ] codestyle - Changes that do not affect the meaning of the code (white-space, formatting, missing semi-colons, etc)
-   [ ] refactor - A code change that neither fixes a bug nor adds a feature
-   [ ] perf - A code change that improves performance
-   [ ] test - Adding missing tests or correcting existing tests
-   [ ] build - Changes that affect the build system or external dependencies (example scopes: gulp, broccoli, npm)
-   [ ] ci - Changes to our CI configuration files and scripts (example scopes: Travis, Circle, BrowserStack, SauceLabs)
-   [ ] chore - Other changes that don't modify src or test files
-   [ ] revert - Reverts a previous commit
-   [ ] content - Changes to content or copy

# Checklist:

-   [x] Able to run pr locally
-   [x] Followed acceptance criteria
-   [x] My code follows the style guidelines of this project
-   [x] I have performed a self-review of my own code
-   [ ] I have made corresponding changes to the documentation if applicable
-   [x] I have added tests that prove my fix is effective or that my feature works

### CODE REVIEW

File: src/pages/api/searchOperators.ts [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+    errors: ErrorInfo[];
+}
+
+export default (req: NextApiRequestWithSession, res: NextApiResponse): void => {
+    const errors: ErrorInfo[] = [];
+
+    try {
+        const { searchText } = req.body;
+
+        if (searchText.length < 4) {
```
Comment (RobCatton) [inline src/pages/api/searchOperators.ts:21]: If this is the only error check we do, could we combine the 2 ifs into 
```
if (searchText.length < 4) {
const errors = [{
                errorMessage: 'Search requires a minimum of three characters',
                id: 'searchText',
                userInput: searchText,
            }]
updateSessionAttribute(req, SEARCH_OPERATOR_ATTRIBUTE, { errors, searchResults: [] });
            redirectTo(res, '/searchOperators');
        }
```
And this way we can render the search they made on the page?

File: src/pages/api/searchOperators.ts [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+                id: 'searchText',
+            });
+        }
+
+        if (errors.length > 0) {
+            updateSessionAttribute(req, SEARCH_OPERATOR_ATTRIBUTE, { errors, searchResults: [] });
+            redirectTo(res, '/searchOperators');
+        }
+
+        updateSessionAttribute(req, SEARCH_OPERATOR_ATTRIBUTE, { errors: [], searchResults: [] });
```
Comment (RobCatton) [inline src/pages/api/searchOperators.ts:33]: Do we need searchResults in the session if its always going to be an empty array?

File: src/pages/searchOperators.tsx [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+        return {
+            props: {
+                errors,
+                searchText: '',
+                operators: [],
+            },
+        };
+    }
+
+    const searchText = searchOperator && searchOperator !== '' ? searchOperator.toString() : '';
```
Comment (RobCatton) [inline src/pages/searchOperators.tsx:142]: could we use our `removeExcessWhiteSpace` method from the validator file to help ensure we get only the characters?

File: src/pages/api/searchOperators.ts (hunk 1 above)
Comment (RobCatton) [inline src/pages/api/searchOperators.ts:21]: I think this is meant to be `.length < 3`

File: src/pages/searchOperators.tsx [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+        <CsrfForm action="/api/searchOperators" method="post" csrfToken={csrfToken}>
+            <>
+                <div className="govuk-form-group">
+                    <h1 className="govuk-label-wrapper">
+                        <label className="govuk-label govuk-label--l" htmlFor="event-name">
+                            Search for the operators that the ticket covers
+                        </label>
+                    </h1>
+                    <FormElementWrapper errors={errors} errorId="searchText" errorClass="govuk-input--error">
+                        <>
```
Comment (RobCatton) [inline src/pages/searchOperators.tsx:39]: I think this div is being rendered on the page as a class-less div, unsure if its this bit of html to blame however. One slight thing - would be be able to add a very small (~10-15px) margin top to the search bar and the button? They just look a little snug to the title is all.

File: tests/pages/api/searchOperators.test.ts [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+    afterEach(() => {
+        jest.resetAllMocks();
+    });
+
+    const updateSessionAttributeSpy = jest.spyOn(session, 'updateSessionAttribute');
+
+    it('should error if the search text if entered is less than 3', () => {
+        const { req, res } = getMockRequestAndResponse({
+            body: {
+                searchText: 'abc',
```
Comment (RobCatton) [inline tests/pages/api/searchOperators.test.ts:18]: This test I think proves that its not working.

File: src/pages/searchOperators.tsx [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+    searchText,
+}: SearchOperatorProps & CustomAppProps): ReactElement => (
+    <TwoThirdsLayout title={title} description={description}>
+        <ErrorSummary errors={errors} />
+        <CsrfForm action="/api/searchOperators" method="post" csrfToken={csrfToken}>
+            <>
+                <div className="govuk-form-group">
+                    <h1 className="govuk-label-wrapper">
+                        <label className="govuk-label govuk-label--l" htmlFor="event-name">
+                            Search for the operators that the ticket covers
```
Comment (RobCatton) [inline src/pages/searchOperators.tsx:35]: I think we're missing either some hint text or a label for the search bar - according to the design, it has Operator Name above the search box.

File: src/pages/searchOperators.tsx [hunk 7]
```diff
@@ -0,0 +1,162 @@
...
+                        </label>
+                    </h1>
+                    <FormElementWrapper errors={errors} errorId="searchText" errorClass="govuk-input--error">
+                        <>
+                            <input
+                                className="govuk-input govuk-!-width-three-quarters"
+                                id="searchText"
+                                name="searchText"
```
Comment (RobCatton) [inline src/pages/searchOperators.tsx:43]: Unsure what's the reason, but we're missing the red line on the error summary render.

File: src/pages/searchOperators.tsx [hunk 8]
```diff
@@ -0,0 +1,162 @@
...
+
+    let operators: OperatorNameType[];
+
+    if (searchText !== '') {
+        const errors: ErrorInfo[] = [];
+
+        operators = await getSearchOperators(searchText, nocCode);
+        if (operators.length === 0) {
+            errors.push({
+                errorMessage: 'No operators found',
```
Comment (RobCatton) [inline src/pages/searchOperators.tsx:152]: Spoke to harry about the error message, could we change it to: 

" No operators found for: {their search} . Try another search term. "

File: src/pages/api/searchOperators.ts (hunk 1 above)
Comment (tejpowar) [reply src/pages/api/searchOperators.ts:21]: can be <= 3 as well so will change

Comment (tejpowar) [reply src/pages/api/searchOperators.ts:21]: i did it so in the next part of the work if we need to do more error handling like at least one checkbox checked etc.

So might be worth leaving it in?

File: src/pages/api/searchOperators.ts (hunk 2 above)
Comment (tejpowar) [reply src/pages/api/searchOperators.ts:33]: removed and if needed for future work can be added

File: tests/pages/api/searchOperators.test.ts (hunk 5 above)
Comment (tejpowar) [reply tests/pages/api/searchOperators.test.ts:18]: changed logic so should

File: src/pages/api/searchOperators.ts [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+export interface SearchOperatorsWithErrors {
+    errors: ErrorInfo[];
+}
+
+export default (req: NextApiRequestWithSession, res: NextApiResponse): void => {
+    const errors: ErrorInfo[] = [];
+
+    try {
+        const { searchText } = req.body;
```
Comment (RobCatton) [inline src/pages/api/searchOperators.ts:15]: We have to clear whitespace on this, otherwise '   ' is a valid search.

**Review by RobCatton — APPROVED** [review]

**Pull request merged**